File: api/contrib/send/send.php
<?php
/*
 * Copyright (C) 2017-2025 CRLibre <https://crlibre.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

function send()
{
    $datos = null;
    $apiTo  = params_get("client_id");
    $url    = ($apiTo == 'api-stag' ? "https://api-sandbox.comprobanteselectronicos.go.cr/recepcion/v1/recepcion/" : ($apiTo == 'api-prod' ? "https://api.comprobanteselectronicos.go.cr/recepcion/v1/recepcion/" : null)); 

    if (params_get("recp_tipoIdentificacion") == "" or params_get("recp_numeroIdentificacion") == "") 
    {
        if (params_get('callbackUrl') == "")
        {
            $datos = array(
                'clave'             => params_get('clave'),
                'fecha'             => params_get("fecha"),
                'emisor'            => array(
                                        'tipoIdentificacion'    => params_get("emi_tipoIdentificacion"),
                                        'numeroIdentificacion'  => params_get("emi_numeroIdentificacion")
                ),
                'comprobanteXml'    => params_get("comprobanteXml")
            );
        }
        else
        {
            $datos = array(
                'clave'             => params_get('clave'),
                'fecha'             => params_get("fecha"),
                'emisor'            => array(
                                        'tipoIdentificacion'    => params_get("emi_tipoIdentificacion"),
                                        'numeroIdentificacion'  => params_get("emi_numeroIdentificacion")
                ),
                'comprobanteXml'    => params_get("comprobanteXml"),
                'callbackUrl'       => params_get('callbackUrl')
            );
        }
    }
    else
    {
        if (params_get('callbackUrl') == "")
        {
            $datos = array(
                'clave'         => params_get('clave'),
                'fecha'         => params_get("fecha"),
                'emisor'        => array(
                                    'tipoIdentificacion'    => params_get("emi_tipoIdentificacion"),
                                    'numeroIdentificacion'  => params_get("emi_numeroIdentificacion")
                ),
                'receptor'      => array(
                                    'tipoIdentificacion'    => params_get("recp_tipoIdentificacion"),
                                    'numeroIdentificacion'  => params_get("recp_numeroIdentificacion")
                ),
                'comprobanteXml' => params_get("comprobanteXml")
            );
        } else {
            $datos = array(
                'clave'     => params_get('clave'),
                'fecha'     => params_get("fecha"),
                'emisor'    => array(
                                'tipoIdentificacion'    => params_get("emi_tipoIdentificacion"),
                                'numeroIdentificacion'  => params_get("emi_numeroIdentificacion")
                ),
                'receptor'  => array(
                                'tipoIdentificacion'    => params_get("recp_tipoIdentificacion"),
                                'numeroIdentificacion'  => params_get("recp_numeroIdentificacion")
                ),
                'comprobanteXml'    => params_get("comprobanteXml"),
                'callbackUrl'       => params_get('callbackUrl')
            );
        }
    }

    //$datosJ= http_build_query($datos);
    $mensaje = json_encode($datos);
    grace_debug("JSON:" . $mensaje);

    $header = array(
        'Authorization: bearer ' . params_get('token'),
        'Content-Type: application/json'
    );

    $curl = curl_init($url);
    curl_setopt($curl, CURLOPT_HEADER, true);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $header);
    curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "POST");
    curl_setopt($curl, CURLOPT_POSTFIELDS, $mensaje);
    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
    $respuesta = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);
    $err = curl_error($curl);
    curl_close($curl);
    if ($err)
    {
        $arrayResp = array(
            "Status"    => $status,
            "to"        => $apiTo,
            "text"      => $err
        );
        return $arrayResp;
    }
    else
    {
        $arrayResp = array(
            "Status"    => $status,
            "text"      => explode("\n", $respuesta)
        );
        return $arrayResp;
    }
}

function sendMensaje()
{
    $apiTo  = params_get("client_id");
    $url    = ($apiTo == 'api-stag' ? "https://api-sandbox.comprobanteselectronicos.go.cr/recepcion/v1/recepcion/" : ($apiTo == 'api-prod' ? "https://api.comprobanteselectronicos.go.cr/recepcion/v1/recepcion/" : null));

    $datos = array(
        'clave'     => params_get('clave'),
        'fecha'     => params_get("fecha"),
        'emisor'    => array(
                        'tipoIdentificacion'    => params_get("emi_tipoIdentificacion"),
                        'numeroIdentificacion'  => params_get("emi_numeroIdentificacion")
        ),
        'receptor'  => array(
                        'tipoIdentificacion'    => params_get("recp_tipoIdentificacion"),
                        'numeroIdentificacion'  => params_get("recp_numeroIdentificacion")
        ),
        'consecutivoReceptor'   => str_pad(params_get("consecutivoReceptor"), 20, "0", STR_PAD_LEFT),
        'comprobanteXml'        => params_get("comprobanteXml")
    );

    //$datosJ= http_build_query($datos);

    $mensaje = json_encode($datos);

    $header = array(
        'Authorization: bearer ' . params_get('token'),
        'Content-Type: application/json'
    );

    $curl = curl_init($url);
    curl_setopt($curl, CURLOPT_HEADER, true);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $header);
    curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "POST");
    curl_setopt($curl, CURLOPT_POSTFIELDS, $mensaje);
    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
    $respuesta = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);
    $err = curl_error($curl);
    curl_close($curl);
    if ($err)
    {
        $arrayResp = array(
            "Status"    => $status,
            "to"        => $apiTo,
            "text"      => $err
        );
        return $arrayResp;
    }
    else
    {
        $arrayResp = array(
            "Status"    => $status,
            "text"      => explode("\n", $respuesta)
        );
        return $arrayResp;
    }
}

function sendTE()
{
    $apiTo = params_get("client_id");
    $url = ($apiTo == 'api-stag' ? "https://api-sandbox.comprobanteselectronicos.go.cr/recepcion/v1/recepcion/" : ($apiTo == 'api-prod' ? "https://api.comprobanteselectronicos.go.cr/recepcion/v1/recepcion/" : null));

    $datos = array(
        'clave'     => params_get('clave'),
        'fecha'     => params_get("fecha"),
        'emisor'    => array(
                        'tipoIdentificacion'    => params_get("emi_tipoIdentificacion"),
                        'numeroIdentificacion'  => params_get("emi_numeroIdentificacion")
        ),
        'comprobanteXml' => params_get("comprobanteXml")
    );

    //$datosJ= http_build_query($datos);

    $mensaje = json_encode($datos);

    $header = array(
        'Authorization: bearer ' . params_get('token'),
        'Content-Type: application/json'
    );

    $curl = curl_init($url);
    curl_setopt($curl, CURLOPT_HEADER, true);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $header);
    curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "POST");
    curl_setopt($curl, CURLOPT_POSTFIELDS, $mensaje);
    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
    $respuesta = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);
    $err = curl_error($curl);
    curl_close($curl);
    if ($err)
    {
        $arrayResp = array(
            "Status"    => $status,
            "to"        => $apiTo,
            "text"      => $err
        );
        return $arrayResp;
    }
    else
    {
        $arrayResp = array(
            "Status"    => $status,
            "text"      => explode("\n", $respuesta)
        );
        return $arrayResp;
    }
}

?>

File: migration/python-api/python_api/handlers/send.py
from __future__ import annotations

import json as jsonlib

import httpx
from fastapi import Request
from fastapi.responses import JSONResponse

from python_api.responses import tools_reply_compatible


def _recepcion_url(client_id: str) -> str | None:
    if client_id == "api-stag":
        return "https://api-sandbox.comprobanteselectronicos.go.cr/recepcion/v1/recepcion/"
    if client_id == "api-prod":
        return "https://api.comprobanteselectronicos.go.cr/recepcion/v1/recepcion/"
    return None


def _raw_http_lines(response: httpx.Response) -> list[str]:
    version = response.http_version or "1.1"
    status_line = f"HTTP/{version} {response.status_code} {response.reason_phrase}\r"
    header_lines = [f"{k}: {v}\r" for k, v in response.headers.items()]
    return [status_line, *header_lines, "\r", response.text]


async def _post_to_hacienda(
    *,
    api_to: str,
    token: str,
    message: str,
) -> JSONResponse:
    url = _recepcion_url(api_to)
    if url is None:
        return tools_reply_compatible({"Status": 0, "to": api_to, "text": "No URL set!"})

    headers = {
        "Authorization": f"bearer {token}",
        "Content-Type": "application/json",
    }

    try:
        async with httpx.AsyncClient(timeout=30, verify=False) as client:
            response = await client.post(url, headers=headers, content=message.encode("utf-8"))
    except Exception as exc:
        return tools_reply_compatible({"Status": 0, "to": api_to, "text": str(exc)})

    return tools_reply_compatible({"Status": response.status_code, "text": _raw_http_lines(response)})


async def json(_: Request, params: dict[str, str]) -> JSONResponse:
    payload: dict[str, object] = {
        "clave": params.get("clave", ""),
        "fecha": params.get("fecha", ""),
        "emisor": {
            "tipoIdentificacion": params.get("emi_tipoIdentificacion", ""),
            "numeroIdentificacion": params.get("emi_numeroIdentificacion", ""),
        },
        "receptor": {
            "tipoIdentificacion": params.get("recp_tipoIdentificacion", ""),
            "numeroIdentificacion": params.get("recp_numeroIdentificacion", ""),
        },
        "comprobanteXml": params.get("comprobanteXml", ""),
        "callbackUrl": params.get("callbackUrl", ""),
    }

    if str(params.get("callbackUrl", "")) == "":
        payload.pop("callbackUrl", None)
    if str(params.get("recp_tipoIdentificacion", "")) == "" or str(params.get("recp_numeroIdentificacion", "")) == "":
        payload.pop("receptor", None)

    message = jsonlib.dumps(payload, ensure_ascii=False)
    return await _post_to_hacienda(
        api_to=str(params.get("client_id", "")),
        token=str(params.get("token", "")),
        message=message,
    )


async def sendMensaje(_: Request, params: dict[str, str]) -> JSONResponse:
    consecutivo_receptor = str(params.get("consecutivoReceptor", "")).zfill(20)

    payload = {
        "clave": params.get("clave", ""),
        "fecha": params.get("fecha", ""),
        "emisor": {
            "tipoIdentificacion": params.get("emi_tipoIdentificacion", ""),
            "numeroIdentificacion": params.get("emi_numeroIdentificacion", ""),
        },
        "receptor": {
            "tipoIdentificacion": params.get("recp_tipoIdentificacion", ""),
            "numeroIdentificacion": params.get("recp_numeroIdentificacion", ""),
        },
        "consecutivoReceptor": consecutivo_receptor,
        "comprobanteXml": params.get("comprobanteXml", ""),
    }

    message = jsonlib.dumps(payload, ensure_ascii=False)
    return await _post_to_hacienda(
        api_to=str(params.get("client_id", "")),
        token=str(params.get("token", "")),
        message=message,
    )


async def sendTE(_: Request, params: dict[str, str]) -> JSONResponse:
    payload = {
        "clave": params.get("clave", ""),
        "fecha": params.get("fecha", ""),
        "emisor": {
            "tipoIdentificacion": params.get("emi_tipoIdentificacion", ""),
            "numeroIdentificacion": params.get("emi_numeroIdentificacion", ""),
        },
        "comprobanteXml": params.get("comprobanteXml", ""),
    }

    message = jsonlib.dumps(payload, ensure_ascii=False)
    return await _post_to_hacienda(
        api_to=str(params.get("client_id", "")),
        token=str(params.get("token", "")),
        message=message,
    )
